client/db.go

```diff
@@ -0,0 +1,473 @@
+		return err
+	}
+	if err := db.kv.Run(b.calls...); err != nil {
+		return err
+	}
+	return b.fillResults()
+}
+
+// Tx ...
+func (db *DB) Tx(retryable func(tx *Tx) error) error {
+	return db.kv.RunTransaction(nil, func(txn *Txn) error {
+		tx := &Tx{txn}
+		return retryable(tx)
+	})
+}
+
+// Tx ...
+type Tx struct {
+	txn *Txn
+}
+
+// Get ...
+func (tx *Tx) Get(args ...interface{}) Result {
+	return get(tx, args...)
+}
+
+// Put ...
+func (tx *Tx) Put(key, value interface{}) error {
+	return put(tx, key, value)
+}
+
+// Inc ...
+func (tx *Tx) Inc(key interface{}, value int64) Result {
+	return inc(tx, key, value)
+}
+
+// Scan ...
+func (tx *Tx) Scan(begin, end interface{}, maxRows int64) Result {
+	return scan(tx, begin, end, maxRows)
+}
+
+// Del ...
+func (tx *Tx) Del(args ...interface{}) error {
+	return del(tx, args...)
+}
+
+// Run ...
+func (tx *Tx) Run(b *Batch) error {
+	if err := b.prepare(); err != nil {
+		return err
+	}
+	if err := tx.txn.Run(b.calls...); err != nil {
+		return err
+	}
+	return b.fillResults()
+}
+
+// Commit ...
+func (tx *Tx) Commit(b *Batch) error {
+	args := &proto.EndTransactionRequest{Commit: true}
+	reply := &proto.EndTransactionResponse{}
+	b.calls = append(b.calls, Call{Args: args, Reply: reply})
+	b.initResult(1, 0, nil)
+	return tx.Run(b)
+}
+
+// Batch ...
+type Batch struct {
+	Results []Result
+	calls   []Call
+	// TODO(pmattis): Use fixed size arrays for Result and KeyValue to
+	// avoid allocations in the common case.
+	// resultsBuf [8]Result
+	// rowsBuf [8]KeyValue
+}
+
+func (b *Batch) prepare() error {
+	for _, r := range b.Results {
+		if err := r.Err; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *Batch) initResult(calls, numRows int, err error) {
+	r := Result{calls: calls, Err: err}
+	if numRows > 0 {
+		r.Rows = make([]KeyValue, numRows)
+	}
+	b.Results = append(b.Results, r)
+}
+
+func (b *Batch) fillResults() error {
+	offset := 0
+	for i := range b.Results {
+		result := &b.Results[i]
+
+		for k := 0; k < result.calls; k++ {
+			call := b.calls[offset+k]
+
+			switch t := call.Reply.(type) {
+			case *proto.GetResponse:
+				row := &result.Rows[k]
+				row.key = []byte(call.Args.(*proto.GetRequest).Key)
+				row.setValue(t.Value)
+			case *proto.PutResponse:
+				row := &result.Rows[k]
+				row.key = []byte(call.Args.(*proto.PutRequest).Key)
+				row.setValue(&call.Args.(*proto.PutRequest).Value)
+			case *proto.IncrementResponse:
+				row := &result.Rows[k]
+				row.key = []byte(call.Args.(*proto.IncrementRequest).Key)
+				row.value = &t.NewValue
+			case *proto.ScanResponse:
+				for _, kv := range t.Rows {
+					var row KeyValue
+					row.key = kv.Key
+					row.setValue(&kv.Value)
+					result.Rows = append(result.Rows, row)
+				}
+			case *proto.DeleteResponse:
+				row := &result.Rows[k]
+				row.key = []byte(call.Args.(*proto.DeleteRequest).Key)
+			}
+		}
+		offset += result.calls
+	}
+	return nil
+}
+
+// Get ...
+func (b *Batch) Get(args ...interface{}) {
```

Comment:
Batch could be more useful if it returned itself, then you could run

`err := b.Get(...).Put(...).Scan(...).Run()`